src/dotnet/CommandLine/CommandLineApplication.cs

```diff
@@ -334,7 +340,15 @@ public void ShowHelp(string commandName = null)
             for (var cmd = this; cmd != null; cmd = cmd.Parent)
             {
                 cmd.IsShowingInformation = true;
-                headerBuilder.Insert(6, string.Format(" {0}", cmd.Name));
+                if (cmd != this && cmd.Arguments.Any())
```

Comment:
Any chance we can break the pieces of this method into smaller, aptly named methods and invoke them here? It is somewhat hard to follow what exactly is going on in this method because there is very little abstraction.